torch/distributed/tensor/parallel/style.py

```diff
@@ -1,652 +1,357 @@
+            # colwise shard weight/bias to Shard(0), weight be Shard(0)
+            # means Colwise as Linear is input * weight^T + bias, where
+            # weight would become Shard(1)
+            for name, param in module.named_parameters():
+                dist_param = nn.Parameter(
+                    distribute_tensor(param, device_mesh, [Shard(0)])
+                )
+                module.register_parameter(name, dist_param)
+        elif isinstance(module, nn.Embedding):
+            # colwise shard embedding.weight is straight forward as Shard(1)
+            for name, param in module.named_parameters():
+                dist_param = nn.Parameter(
+                    distribute_tensor(param, device_mesh, [Shard(1)])
+                )
+                module.register_parameter(name, dist_param)
+        else:
+            raise NotImplementedError(
+                "ColwiseParallel only supports nn.Linear"
+                f"and nn.Embedding for now, but found {type(module)}!"
             )
 
-        super().__init__(
-            input_layouts=input_layouts,
-            output_layouts=output_layouts,
-            use_local_output=use_local_output,
-            _prepare_input=prepare_input_fn,
-            _prepare_output=prepare_output_fn,
-        )
-
-    @staticmethod
-    def _prepare_input_fn(input_layouts, output_layouts, inputs, device_mesh=None):
-        return (
-            _redistribute_per_both_layouts(
-                inputs[0], input_layouts, output_layouts, device_mesh
-            ),
+    def _prepare_output_fn(self, outputs, device_mesh):
+        # outputs is a shard on last dimension DTensor, i.e. Shard(-1)
+        if outputs.placements != self.desired_output_layouts:
+            outputs = outputs.redistribute(placements=self.desired_output_layouts)
+        # back to local tensor
+        return outputs.to_local() if self.use_local_output else outputs
+
+    def apply(self, module: nn.Module, device_mesh: DeviceMesh) -> nn.Module:
+        return distribute_module(
+            module,
+            device_mesh,
+            self._partition_fn,
+            self._prepare_input_fn,
+            self._prepare_output_fn
         )
 
 
-class ColwiseParallel(ParallelStyle):
+class RowwiseParallel(ParallelStyle):
     """
-    Partition the column of a tensor or module.
-
-    We assume the input to be a replicated :class:`DTensor` and output to be a sharded :class:`torch.Tensor`.
-
-    Args:
-        input_layouts (Union[Placement, Tuple[Placement, ...]]):
-            The layout of input tensor(s) which DTensor will be created upon.
-        output_layouts (Union[Placement, Tuple[Placement, ...]]):
-            The layout of input tensor(s) which created DTensor will be redistributed to.
-        use_local_output (bool):
-            Whether to convert the DTensor to local :class:`torch.Tensor`.
-
+    Partition the row of a nn.Module and allow the parallelized module to run rowwise sharded computation.
```

Comment:
similar to my comment for Colwise, and after i read more of the docs/code, i think something like this makes sense

Partition a compatible nn.Module in a row-wise fashion.  Currently supports nn.Linear only.